### ISSUE TITLE
Roadmap: verifiable-intelligence receipts and statement-bound agent steps

### ISSUE THREAD

**Issue by omarespejel** [omarespejel/llm-provable-computer#308]
## Purpose

This issue tracks the next research program after the Tablero release: moving from **STARK-zkML** toward **verifiable intelligence**.

Working thesis:

> Verifiable intelligence is not only proving that a model inference ran. It is proving that an intelligent system followed a committed computation, context, policy, and state transition.

The goal is to make this repo's next line of work explicit before we scatter it across one-off benchmark, prover, and paper issues.

## Current read

Tablero is useful for this agenda, but only in the correct role.

Tablero should **not** be positioned as the whole verifiable-intelligence system. It does not prove model semantics, tool truthfulness, memory correctness, policy compliance, or agent reasoning quality.

Tablero **does** provide a useful typed verifier boundary once those facts are proven or attested elsewhere:

```text
agent/model/tool/memory/policy evidence
        -> statement-bound typed receipt
        -> Tablero boundary
        -> compact verifier / settlement / audit object
```

In other words: Tablero is the receipt/boundary layer, not the intelligence layer.

## Why this matters

The external landscape is already crowded on isolated zkML and generic zkVM proving:

- EZKL-style systems own the developer-friendly ONNX/Halo2 lane.
- zkLLM / zkGPT / NANOZK-style systems are strong on specialized LLM-layer proof systems.
- Nexus zkMCP-style work is framing the broader problem as proving model execution context: what model ran, with which input, under which configuration.
- Artemis-style commit-and-prove work suggests commitment consistency can become a dominant zkML bottleneck.
- Percepta-style "LLMs as computers" work, whether or not the current public claims hold, pushes the narrative from model inference toward computation inside/around intelligent systems.

The differentiated lane for this repo should be:

> Transparent, statement-bound STARK receipts for AI/agent computation, with Tablero as the typed boundary that makes those receipts composable and replay-safe.

## Core object: agent-step receipt

A verifiable-intelligence step should eventually bind an agent state transition:

```text
(state_t, observation_t, policy_t, model_t, tools_t)
        -> (state_{t+1}, action_t, transcript_t)
```

The receipt should domain-separate and bind at least:

- prior state commitment
- observation/input commitment
- model identity and model/weights commitment
- quantization/config commitment when model inference is involved
- tool outputs or oracle attestations consumed by the step
- policy/rule-set commitment
- action/output commitment
- next-state commitment
- runtime/backend/proof-system version
- transcript hash
- dependency-drop manifest: which replay surfaces are being replaced by commitments or typed sub-receipts

## Track A: finish Tablero release and close verifier-boundary hardening

Immediate PR sequence:

1. Finish and merge PR #307 after review discipline.
   - Complete remaining CodeRabbit/Qodo actionable items.
   - Preserve the 5-minute quiet-window gate after latest AI-reviewer activity.
   - Regenerate evidence from final code.
   - Re-run local release/preflight gates.

2. Close or update #291 from the merged microprofile evidence.
   - The target outcome is not a bigger headline number.
   - The target outcome is confirmation that typed-boundary verification does not hide replay-linear work.

3. Keep #284 / #258 (SNIP-36) deferred.
   - Useful later for deployment grounding.
   - Not a blocker for the verifiable-intelligence research track.

## Track B: verifiable-intelligence positioning and literature map

Next PR:

- Add a paper/research note mapping the landscape:
  - EZKL
  - zkLLM
  - zkGPT
  - NANOZK
  - Nexus zkMCP
  - Artemis / commit-and-prove zkML
  - Percepta / LLMs-as-computers framing
  - BitSage/Obelyzk as ecosystem signal, not the main competitor

Deliverable:

- A concise `docs/engineering/` or `docs/research/` note explaining why our next lane is not "another zkML prover" but statement-bound verifiable intelligence receipts over transparent STARKs.

Non-goal:

- Do not claim we beat SNARK zkML systems.
- Do not claim Tablero proves intelligence.

## Track C: statement-bound transformer proof

This is the core zkML/prover track.

Goal:

> A Stwo-native proof for a small but real transformer primitive/block whose public statement cannot be relabeled.

Minimum target:

- quantized linear/RMSNorm or a similarly meaningful transformer primitive
- Stwo-native AIR/proof path, not Cairo execution as the primary proving substrate
- statement binds:
  - model/weights commitment
  - input commitment
  - output commitment
  - quantization config
  - operator/block identifier
  - backend/proof version
  - verifier domain

Good target:

- `d=32` transformer block or meaningful sub-block
- reproducible commands
- checked-in evidence
- mutation tests that reject every relabel attempt

Excellent target:

- `d=64` or `d=128` block/sub-block
- scaling curve
- matched EZKL comparator on same primitive/dimensions
- Tablero receipt wrapping the accepted statement

## Track D: adversarial statement/relabeling harness

This should be a first-class engineering artifact, not an afterthought.

Build a generic mutation harness for accepted receipts. Each mutation must reject:

- model ID changed
- weights commitment changed
- input commitment changed
- output commitment changed
- quantization config changed
- policy hash changed
- tool output hash changed
- state transition hash changed
- backend/proof version changed
- proof replayed under another statement domain

This becomes the repo's strongest "we are not fooling ourselves" artifact.

Potential standalone contribution:

> A zkML/verifiable-intelligence relabeling benchmark suite.

This could be useful even beyond our own prover.

## Track E: matched comparators

Add comparators only after the statement spec is fixed.

Priority order:

1. EZKL on the same primitive/dimensions.
2. NANOZK/zkGPT/zkLLM paper comparison where exact reproduction is not feasible.
3. Nexus zkMCP as conceptual/context-receipt comparator, not a transformer-prover comparator.
4. BitSage/Obelyzk only as an ecosystem deployment signal unless their current code/deployed verifier passes independent relabeling and source/class-hash retesting.

Comparison axes:

- prover time
- verifier time
- proof size
- setup assumptions
- statement-binding coverage
- reproducibility
- auditability of constraints/AIR/circuit
- deployment surface

## Track F: proof-carrying KV-cache / agent state transitions

High-upside exploratory idea.

Autoregressive inference is naturally stateful. Instead of treating LLM verification as one monolithic transcript, define typed transition receipts:

```text
kv_state_t, token_t, model_commitment
        -> kv_state_{t+1}, logits_or_action_t
```

Why it may matter:

- Fits agent execution better than one-shot inference proofs.
- Makes incremental verification possible.
- Aligns with Tablero's typed-boundary pattern.
- Gives a route to proof-carrying memory/state for agents.

Exploratory deliverable:

- Short design note defining a proof-carrying KV-cache receipt.
- Decide whether the first feasible implementation is a tiny transformer block, synthetic KV transition, or receipt-only mock with future prover hooks.

## Track G: tool, policy, and memory proof adapters

Verifiable intelligence will need heterogeneous evidence, not only model proofs.

Possible sub-receipts:

- model inference proof
- tool-output attestation or proof
- memory/state transition proof
- policy-compliance proof
- transcript commitment
- action/output commitment

First step:

- Define the typed receipt schema and domain separation.
- Allow some fields to be `attested` rather than `proved`, but require that the trust class is explicit.

This prevents false claims like "fully proven agent" when some parts are only signed or oracle-attested.

## Track H: audit packet / formalization

Before stronger public claims:

- theorem-style statement for the agent-step receipt binding property
- threat model: relabeling, replay, stale tool output, stale state, policy substitution, backend downgrade
- mutation-test matrix
- reproducibility commands
- exact evidence handles
- known non-goals and unproven surfaces

This should be prepared for a STARK/proof-systems specialist review later.

## Proposed next PR order

1. **PR #307 completion**: Tablero boundary-binding microprofile.
2. **Landscape/positioning note**: verifiable intelligence vs zkML vs Tablero.
3. **Agent-step receipt spec**: schema, commitments, domain separation, non-goals.
4. **Statement-bound transformer primitive spec**: exact public statement and AIR/prover target.
5. **Relabeling mutation harness**: generic receipt mutation framework.
6. **Stwo-native primitive feasibility PR**: first transformer primitive proof or documented no-go.
7. **EZKL comparator PR**: same primitive/dimension if feasible.
8. **Tablero receipt integration PR**: wrap the primitive proof as a typed receipt boundary.
9. **Proof-carrying KV-cache design PR**: high-upside next-paper branch.

## Open research questions

- Is the first practical Stwo-native target RMSNorm, quantized linear/matmul, MLP/SwiGLU, or attention projection?
- Can the typed receipt be made generic enough for tool/memory/policy evidence without becoming vague?
- What is the right trust taxonomy: proved, attested, replayed, omitted, or dependency-dropped?
- Is proof-carrying KV-cache actually implementable in a small enough first artifact?
- Does Tablero remain a clean boundary abstraction once receipts include heterogeneous non-STARK evidence?
- Can we build a useful public relabeling benchmark suite that other zkML/verifiable-AI systems can run?

## Non-goals for now

- Do not prioritize SNIP-36 before the statement and receipt model is clean.
- Do not chase larger Tablero frontier ratios as the main research program.
- Do not claim backend independence without second-backend evidence.
- Do not claim full verifiable intelligence from a model-only proof.
- Do not make BitSage the central competitor; treat it as ecosystem signal and diligence input.

## Success criteria

Near-term success:

- Tablero release finished and microprofile hardening merged.
- A clear verifiable-intelligence receipt spec exists.
- Relabeling/mutation threat model is explicit.

Medium-term success:

- One Stwo-native transformer primitive proof exists.
- The accepted statement is mutation-tested and cannot be relabeled.
- A Tablero-style receipt wraps the accepted proof.
- One matched EZKL comparator exists.

High-upside success:

- Proof-carrying agent/KV-cache transition receipts become a distinct research result.
- The repo can claim a path from transparent STARK zkML to verifiable agent state transitions without pretending every part is already solved.

**Comment by omarespejel** [comment]
## Exploration triage update (April 29, 2026)

Split the roadmap into focused follow-up issues so the next work can proceed PR-by-PR:

- #309 — agent-step receipt spec. This is the best immediate PR after #307 because it creates the exact object that makes Tablero useful for verifiable intelligence without overclaiming that Tablero proves intelligence.
- #310 — Stwo-native statement-bound transformer primitive feasibility. This is the zkAI/prover track underneath the broader agent roadmap.
- #311 — receipt mutation / relabeling rejection harness. This should follow the spec so it targets a stable statement surface.

Current sequencing read:

1. Finish #307 with the review-gated merge discipline.
2. Land #309 as a docs/spec PR.
3. Use #309 to sharpen #310 and #311 before implementation.
4. Keep SNIP-36 parked until the receipt and statement models are stable.

Important claim guardrail: the verifiable-intelligence track should say "Tablero is the typed receipt/boundary layer" and not "Tablero proves agents" or "Tablero proves intelligence."

**Comment by omarespejel** [comment]
## Update from PR #323

PR #323 moves the roadmap forward in two ways:

1. Adds `zkAIStatementReceiptV1` as the model/proof subreceipt under `AgentStepReceiptV1`.
2. Adds a second external adapter, Circom/snarkjs Groth16, after EZKL.

New external-adapter result:

- EZKL: proof-only rejects `1 / 7`; statement envelope rejects `7 / 7`.
- snarkjs: proof-only rejects `1 / 9`; statement envelope rejects `9 / 9`.

Interpretation stays narrow and useful: proof validity and statement binding are distinct verifier layers. This supports the verifiable-intelligence direction without claiming that Tablero proves agents or that external proof systems are weak.

**Comment by omarespejel** [comment]
Update after PR #323 hardening pass: the snarkjs adapter mutation suite was expanded from 9 to 14 surfaces, including `statement_commitment`, setup, version, input artifact hash, and verification-key file hash.

Current checked result:

- raw `snarkjs-proof-only`: baseline accepted, rejected `1 / 14` relabeling mutations
- `snarkjs-statement-envelope`: baseline accepted, rejected `14 / 14` relabeling mutations

The evidence files and PR body now use the hardened `14`-mutation suite.

**Comment by omarespejel** [comment]
Added #324 as the focused research thread for the external-adapter lesson: a proof is not a statement, and zkAI/verifiable-intelligence systems need statement receipts that bind proof validity to model/input/output/policy/verifier context. This keeps the ecosystem critique separate from the broader roadmap and from the Stwo-native primitive work in #310.

**Comment by omarespejel** [comment]
Roadmap update after PR #327:

- `#310` is closed with a native Stwo GO result.
- The receipt pattern now has external adapters plus a native Stwo primitive adapter.
- Follow-up tracks are split out:
  - `#325`: larger Stwo-native statement-bound transformer block.
  - `#326`: compose a zkAI statement receipt into `AgentStepReceiptV1`.

Near-term recommendation: work `#326` if the goal is verifiable-intelligence composition, or `#325` if the goal is stronger zkML computation evidence.

**Comment by omarespejel** [comment]
Roadmap update after PR #365 (`7ea9ab4ff5641cbfa31fb20893da1dab35726ee3` on `main`).

New concrete results:
- Range-disciplined activation receipt: `35 / 35` relabeling mutations rejected over source-backed JSTprove/Remainder ReLU scaling evidence.
- Attention/KV transition receipt: `8 / 8` state/relabeling mutations rejected on a tiny source-backed stateful transition fixture.
- Native d64 RMSNorm public-row proof hardening: bounded sqrt inequality moved into Stwo AIR with 17-bit gap decompositions, plus local `rmsnorm_output_row_commitment` recomputation.

Research interpretation:
- The verifiable-intelligence track is now less speculative. We have concrete evidence that model statements need explicit binding for numeric range assumptions and carried state transitions, not just model/input/output labels.
- Still no claim of full zkAI or agent correctness. The stateful attention/KV result is source-backed; the next upgrade is proof-backed/native.

Next best tracks:
1. #358: consume `rmsnorm_output_row_commitment` in the next d64 relation surface.
2. #336: turn the source-backed attention/KV receipt into a proof-backed/native stateful receipt.
3. #342: external d64 adapter control with exact statement semantics and explicit scale/range drift reporting.
